### PR TITLE
First integration and unit tests

### LINKED ISSUE #20: Filesystem testing: Introduce integration testing capabilities
depends on #21

### PR DESCRIPTION
**Pull request by jens1o**
closes #20

### CODE REVIEW

File: src/storage/loader/filesystem.rs [hunk 1]
```diff
@@ -22,3 +22,63 @@ impl Loader for FilesystemLoader {
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::FilesystemLoader;
+    use crate::storage::loader::Loader;
+    use std::sync::Arc;
+
+    #[tokio::test]
+    async fn test_read_access() {
```
Comment (fiji-flo) [inline src/storage/loader/filesystem.rs]: We could  return a `Result` und use `?`

File: src/storage/saver/filesystem.rs [hunk 2]
```diff
@@ -94,16 +93,26 @@ impl Saver for FilesystemSaver {
...
         let file_uuid = Uuid::new_v4().to_simple().to_string();
-        let file_name = format!("{}-{}", bucket, file_uuid);
+        let file_name = format!("tmp-{}", file_uuid);
 
         Box::pin(async move {
+            match fs::create_dir_all(path.clone()).await {
+                Ok(()) => (),
+                // ignore error, as an error is also thrown when the directory already exists
+                // if the error is fatal, the write operation will also fail
+                Err(err) => info!(
```
Comment (fiji-flo) [inline src/storage/saver/filesystem.rs]: Make it `error!` and lowercase, please.

File: src/tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    let avatar_settings = Data::new(AvatarSettings {
+        s3_bucket: String::from("test_avatar_bucket"),
+        retrieve_by_id_path: String::from("/avatar/get/id/"),
+        picture_api_url: String::from("http://localhost"),
+    });
+
+    let cis_client = Data::new(MockCisClient {});
+
+    let time_to_live = ::std::time::Duration::from_secs(60 * 60 * 24);
+    let cache = Data::new(RwLock::new(
```
Comment (fiji-flo) [inline src/tests/mod.rs]: That must be a mutex now.

File: src/tests/mod.rs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+    #[derive(Deserialize, Debug)]
+    struct UuidResponse {
+        pub uuid: String,
+    }
+
+    let res_json: UuidResponse = test::read_response_json(&mut app, req).await;
+
+    // quick sanity check that the uuid is valid
+    assert_eq!(res_json.uuid.len(), 32);
```
Comment (fiji-flo) [inline src/tests/mod.rs]: Just try to create a UUID from that https://docs.rs/uuid/0.8.1/uuid/struct.Uuid.html#method.parse_str

File: src/tests/mod.rs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        display = "public"
+    ));
+
+    let req = test::TestRequest::post()
+        .uri(&format!(
+            "/internal/save/{uuid}?@@testScope@@=authenticated",
+            uuid = res_json.uuid
+        ))
+        .header("Content-type", "application/json")
+        .set_payload(save_payload)
```
Comment (fiji-flo) [inline src/tests/mod.rs]: Just use `set_json` and the `json!` macro

**Review by fiji-flo — CHANGES_REQUESTED** [review]
Reviewed from my phone. I’ll take another look from my laptop. 

Overall great work!

**Comment by jens1o** [comment]
~~Do you have an idea why it is failing on Travis PR and GitHub Workflow, but works fine with the Travis CI push hook and on my local machine? Is it the cache:tm:?~~

File: src/tests/mod.rs (hunk 3 above)
Comment (jens1o) [reply src/tests/mod.rs]: great catch, thanks :)

**Comment by jens1o** [comment]
I messed it up, let me try to squash the commits. :/

File: src/tests/mod.rs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    let req = test::TestRequest::get().uri("/healthz").to_request();
+
+    let res = test::call_service(&mut app, req).await;
+    assert!(res.response().status().is_success());
+
+    Ok(())
+}
+
+#[actix_rt::test]
+#[cfg(feature = "localuserscope")]
```
Comment (fiji-flo) [inline src/tests/mod.rs]: That’s not needed, is it?

Comment (jens1o) [reply src/tests/mod.rs]: Ah, nice catch. Yeah, it is not needed anymore when we supply `--all-features`

**Review by fiji-flo — APPROVED** [review]

**Pull request merged**